tw-tkms-starter/src/main/java/com/transferwise/kafka/tkms/TkmsTopicValidator.java

```diff
@@ -0,0 +1,214 @@
+          } catch (Throwable t) {
+            log.error("Topic validation for '" + topic + "' failed.", t);
+            failures.incrementAndGet();
+          } finally {
+            countDownLatch.countDown();
+            semaphore.release();
+          }
+        });
+      } else {
+        break;
+      }
+    }
+
+    final var timeoutMs = tkmsProperties.getInternals().getTopicPreValidationTimeout().toMillis() - System.currentTimeMillis() + startTimeEpochMs;
+
+    if (!ExceptionUtils.doUnchecked(() -> countDownLatch.await(timeoutMs, TimeUnit.MILLISECONDS))) {
+      tkmsKafkaProducerProvider.closeKafkaProducerForTopicValidation();
+      throw new IllegalStateException("Topic validation is taking too long.");
+    }
+
+    if (failures.get() > 0) {
+      tkmsKafkaProducerProvider.closeKafkaProducerForTopicValidation();
+      throw new IllegalStateException("There were failures with topics validations. Refusing to start.");
+    }
+  }
+
+  @Override
+  public void validate(TkmsShardPartition shardPartition, String topic, Integer partition) {
+    if (tkmsProperties.isUseAdminClientForTopicsValidation()) {
+      validateUsingAdmin(shardPartition, topic, partition);
+    } else {
+      validateUsingProducer(topic);
+    }
+  }
+
+  protected void validateUsingAdmin(TkmsShardPartition shardPartition, String topic, Integer partition) {
+    topicsValidatedDuringInitializationOrNotified.computeIfAbsent(topic, k -> {
+      problemNotifier.notify(shardPartition.getShard(), NotificationType.TOPIC_NOT_VALIDATED_AT_INIT, NotificationLevel.WARN, () ->
+          "Topic '" + topic + "' was not validated during initialization. This can introduce some lag."
+              + " Please specify all the topics this service is using in the Tkms property of 'topics'."
+      );
+      return Boolean.TRUE;
+    });
+
+    var response = topicDescriptionsCache.get(new FetchTopicDescriptionRequest().setTopic(topic));
+
+    if (response == null) {
+      throw new NullPointerException("Could not fetch topic description for topic '" + topic + "'.");
+    }
+
+    if (response.getThrowable() != null) {
+      String message;
+      if (response.getThrowable() instanceof UnknownTopicOrPartitionException) {
+        message = "Topic '" + topic + "' does not exist.";
+      } else {
+        message = "Topic validation for '" + topic + "' failed.";
+      }
+      throw new IllegalStateException(message, response.getThrowable());
+    }
+
+    final var topicDescription = response.getTopicDescription();
+
+    final var aclOperations = topicDescription.authorizedOperations();
+    if (aclOperations == null || aclOperations.isEmpty()) {
+      tkmsMetricsTemplate.registerNoAclOperationsFetched(shardPartition, topic);
+    } else if (!aclOperations.contains(AclOperation.ALL)
+        && !aclOperations.contains(AclOperation.WRITE)
+        && !aclOperations.contains(AclOperation.IDEMPOTENT_WRITE)
```

Comment:
IDEMPOTENT_WRITE is not enough by itself, we need WRITE or ALL too. That is IDEMPOTENT_WRITE does not include WRITE implicitly. We have a catch all rule for all clusters to allow everyone to do I_W: https://github.com/transferwise/kafka-topic-configurations/blob/2392c8d3289af69254bcd5cecd0c8e8770b70428/main/production/acls/cluster/cluster-acls.yml#L25-L27